packages/mui-system/src/createStyled.js

```diff
@@ -217,105 +158,52 @@ export default function createStyled(input = {}) {
       label,
       ...options,
     });
-    const muiStyledResolver = (styleArg, ...expressions) => {
-      const expressionsWithDefaultTheme = expressions
-        ? expressions.map((stylesArg) => {
-            // On the server Emotion doesn't use React.forwardRef for creating components, so the created
-            // component stays as a function. This condition makes sure that we do not interpolate functions
-            // which are basically components used as a selectors.
-            if (typeof stylesArg === 'function' && stylesArg.__emotion_real !== stylesArg) {
-              return (props) =>
-                muiStyledFunctionResolver({ styledArg: stylesArg, props, defaultTheme, themeId });
-            }
-            if (isPlainObject(stylesArg)) {
-              let transformedStylesArg = stylesArg;
-              let styledArgVariants;
-
-              if (stylesArg && stylesArg.variants) {
-                styledArgVariants = stylesArg.variants;
-                delete transformedStylesArg.variants;
-
-                transformedStylesArg = (props) => {
-                  let result = stylesArg;
-                  const variantStyles = variantsResolver(
-                    props,
-                    transformVariants(styledArgVariants),
-                    styledArgVariants,
-                  );
-                  variantStyles.forEach((variantStyle) => {
-                    result = deepmerge(result, variantStyle);
```

Comment:
I just noticed this `deepmerge`. I think we should remove it and let emotion/styled-components handle it. Both of them already support array of styles.

@mnajdova Should we try to remove it? the perf should be a bit better while producing the same result.

Reply:
Yes, it would work but there is a tiny difference in the dev tool

without merging style, you might see:

```css
.hashed-variant2:hover {
  color: red;
}

.hashed-variant1:hover {
  color: green;
}
```

Reply:
I'd not merge so that the logic does not depend on the `deepmerge` but rely on the style engine.

Reply:
Have we tried removing the deepmerge and see if some test are failing?

Reply:
@mnajdova All green, see the change in this https://github.com/mui/material-ui/pull/40690/commits/473e62d554f3767f86c5645b0df82478bdd0b01b